File: leetcode380.py
import random

class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.l = []
        self.dic = {}

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.dic:
            return False
        # 插入list的最后一个位置
        self.dic[val] = len(self.l)
        self.l.append(val)
        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val in self.dic:
            # move the last element to the place idx of the element to delete
            last_element, idx = self.l[-1], self.dic[val]
            self.l[idx], self.dic[last_element] = last_element, idx
            # delete the last element
            self.l.pop()
            del self.dic[val]
            return True
        return False

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        return random.choice(self.l)

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: leetcode/leetcode380.cc
#include <unordered_map>
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <unordered_set>
#include <cmath>

using namespace std;

class RandomizedSet {
private:
    unordered_set<int> us;
    vector<int> vec;
public:
    RandomizedSet() {
        vec.reserve(100);
    }

    bool insert(int val) {
        if (us.count(val)) {
            return false;
        }
        us.insert(val);
        return true;
    }

    bool remove(int val) {
        if (us.count(val)) {
            us.erase(val);
            return true;
        }
        return false;
    }

    int getRandom() {
        int sz = us.size();
        int place = rand() % sz;
        vec.assign(us.begin(), us.end());
        return vec[place];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */